photon-ml/src/main/scala/com/linkedin/photon/ml/data/KeyValueScore.scala

```diff
@@ -63,7 +63,12 @@ protected[ml] class KeyValueScore(val scores: RDD[(Long, Double)]) extends RDDLi
     val addedScores =
       this.scores
         .fullOuterJoin(that.scores)
-        .mapValues { case (thisScore, thatScore) => thisScore.getOrElse(0.0) + thatScore.getOrElse(0.0) }
+        .mapValues { case (thisScore, thatScore) =>
+          require(!thisScore.isEmpty || !thatScore.isEmpty, s"Should not have reached this state. A full outer join " +
+            s"should yield a result that has at least one of $thisScore or $thatScore non-empty")
+          thisScore.getOrElse(thatScore.get.getZeroScoreDatum()) +
+            thatScore.getOrElse(thisScore.get.getZeroScoreDatum())
```

Comment:
That looks much better, but I would suggest we merge that with the code of the `fullOuterJoin` on the previous line, which is (so parallel!):

```scala
this.cogroup(other, partitioner).flatMapValues {
      case (vs, Seq()) => vs.iterator.map(v => (Some(v), None))
      case (Seq(), ws) => ws.iterator.map(w => (None, Some(w)))
      case (vs, ws) => for (v <- vs.iterator; w <- ws.iterator) yield (Some(v), Some(w))
 }
```

So, if we just use a `cogroup` and 3 cases (the 4th cannot happen), we are done, and it should be shorter and faster.